### ISSUE TITLE
Issue HG5: HardenedGuard Ablation Study - Every Component Earns Its Place

### ISSUE THREAD

**Issue by Mati86** [AI-Security-Internships-2026/03-prompt-injection-detection#33]
# Issue HG5: HardenedGuard Ablation Study — Every Component Earns Its Place

## Milestone: HG5 — Ablation Study & Component-Level Analysis

---

## Objective

Prevent the #1 critique of "composite guard" papers:

> "You combined 4 existing ideas and called it new. You didn't prove ANY component individually matters."

This milestone runs a systematic ablation of HardenedGuard. Every module must demonstrate a measurable contribution, or it gets demoted (kept in code but not claimed as a contribution).

The existing `eval_paper.py` already does a version of this on eval_dataset_v2 (raw → +canonicalize → +calibrated → +rules → full). HG5 generalizes it to: (a) both datasets, (b) adds cross-turn ablation, (c) measures LATENCY not just accuracy, and (d) isolates canonicalization overhead explicitly.

---

## 1. Ablation Ladder (All 6 Variants Required)

Train and evaluate each of the following 6 configurations on BOTH eval_v2 AND deepset:

| Variant # | Name | What's included | Interpretation |
|---|---|---|---|
| A | Raw / Unwrapped Guard | Semantic core (PIGuard class) only. No canonicalization, no extra rules, no calibration, no cross-turn. | "Worst case — what any user gets out of the box." |
| B | A + Canonicalization (wrap_guard) | Raw guard + 8-view bounded canonicalization (standalone preprocessor from HG3). | "Solo contribution of canonicalization as a preprocessor." |
| C | B + Threshold Calibration | B + operating threshold tuned on calibration split only. Not test split. | "Contribution of calibration. Many guards' bad numbers come from bad default thresholds." |
| D | C + HardenedGuard Rule Layer | C + the rule-engine components from `hardened_guard.py` (signature-based rules, heuristic detectors, resource checks). | "Marginal contribution of rules." |
| E | D + Cross-Turn Detector (Conversation-Level) | D + `multiturn.py` conversation-level detection on multi-turn conversation splits. | "Marginal contribution of cross-turn persistence detection." |
| F | FULL HardenedGuard (Ours) | E with all components enabled. | "Final system we actually claim." |

- [ ] Variants A through F defined in code (use a config / flag, not 6 separate scripts).
- [ ] Each variant run on eval_v2 AND deepset, on the same held-out test split (shuffled the same way for comparability).
- [ ] Cross-turn (Variant E vs D) only measurable on CONVERSATION datasets. If deepset is single-turn only, run E vs D on eval_v2's multi-turn subset AND on Garak multi-turn if available. Mark with note: "E vs D not measurable on single-turn datasets."

---

## 2. Metrics per Variant (Beyond Accuracy)

For each variant × dataset:
- [ ] Recall, FPR, Precision, Macro-F1, PR-AUC, ROC-AUC (where guard outputs scores, not just labels).
- [ ] **Latency per-sample** (p50, p95, p99, max) — measured in ms on a reference idle single-CPU-core machine (no GPU noise). Aggregated across ≥1000 samples to stabilize.
- [ ] **Peak memory per input** if measurable. Memory matters for production deployment.
- [ ] **Tokens processed per input**: raw token count vs sum-of-all-views token count. If canonicalization expands tokens by 8×, that's a real cost. Measure it and compare.

---

## 3. Isolate Canonicalization Overhead Explicitly

Canonicalizer is the novel preprocessor. Separate its cost from semantic guard cost.

- [ ] Measure: `latency(variant B) - latency(variant A)` = canonicalizer add-on overhead.
- [ ] Report p50/p95 of this overhead on reference hardware.
- [ ] Compare to: end-to-end latency of full variant F. What % of total is canonicalization? (Should be low — < 10% ideally; if higher, investigate bottlenecks.)
- [ ] If canonicalization is 40% of total latency: optimization opportunity → profile in code, identify slowest transform, fix if easy, or document as acceptable trade-off if not.

---

## 4. Calibration Disjoint Check — No Leakage

The plan explicitly requires: "Disjoint calibration set used for threshold selection." Verify it's actually disjoint.

- [ ] Add a leakage test: in `tests/leakage/`, assert that calibration row indices and test row indices have empty intersection.
- [ ] For k-fold: assert calibration fold and test fold are disjoint per run.
- [ ] If any evaluation script accidentally fits the threshold on combined train+test data, fix it BEFORE ablation runs. Do not run ablation on a leaked pipeline.

---

## 5. ROC / PR Curves — Calibrated Operating Points

For each variant that produces scores (not just binary labels):
- [ ] Plot ROC curve on test split of eval_v2; mark the operating point used.
- [ ] Plot PR curve (superior to ROC under class imbalance).
- [ ] For the FULL variant (F): plot its operating point against variants A/B/C on the SAME ROC/PR axes. This figure shows visually: "Full variant dominates raw guard in both dimensions."

---

## 6. Pareto / Security-Performance Trade-Off Figure

Create a single visualization for paper Figure 5 or similar:

**X-axis:** p95 latency per request (ms)
**Y-axis:** Recall on eval_v2 at 1% FPR (hold FPR fixed at ≤ 1% across all variants so comparison is fair)

Plot all 6 variants as points:
```
Variant F (Full Ours)      — top-right ideally
Variant E (D + cross-turn)
Variant D (C + rules)
Variant C (B + calibrated)
Variant B (A + canonicalize)
Variant A (Raw guard)      — bottom-left usually
```

Draw a convex hull around non-dominated points → Pareto frontier.

**Interpretation paragraph drafted for paper:** "The full HardenedGuard is Pareto-optimal, adding 42 percentage points of recall over the raw guard while adding X ms of latency at p95. The largest single contributor is [canonicalization / calibration / rules / cross-turn]."

---

## 7. Non-Contributing Component Handling

After ablations, the student and supervisor check for weak links:

**Case 1 — Component contributes ≥ 2% recall at ≤ 1% FPR, or ≥ 5% F1:** Keep it as a headline contribution. Claim it.

**Case 2 — Component contributes < 1% recall and < 2% F1 and adds > 10% latency:**
- [ ] **DO NOT delete it from code.** It still lives in the repo for future users.
- [ ] **DO demote it from paper claims.** In the paper, write: "Rule layer was evaluated in ablation and contributed less than 1% F1 in our tested settings; we therefore treat canonicalization + calibration + cross-turn as the primary novelty and include the rule layer only as an optional configurable."
- [ ] Record demotion explicitly in an issue comment. Supervisor signs off.

No embarrassment in this. Honesty about ablation results improves paper quality and reduces reviewer pushback.

---

## Constraints

- Do **NOT** tune any hyperparameter (including thresholds) on test split data. Calibration split only.
- Do **NOT** cherry-pick ablation results. If a component hurts, say so.
- Do **NOT** include latency measurements during GPU contention. Use idle machine or isolated core.
- All figures regenerated from CSVs via script (same pattern as KV6 one-click generation).

---

## Definition of Done (9 Checks)

- [ ] All 6 ablation variants run on eval_v2 AND deepset.
- [ ] Recall, FPR, Precision, Macro-F1, ROC-AUC, PR-AUC reported per variant × dataset.
- [ ] Latency per variant measured and reported (p50 / p95 / p99 / max).
- [ ] Canonicalization overhead isolated (difference B - A) and reported.
- [ ] Calibration / test disjointness verified programmatically (dedicated leakage test).
- [ ] ROC and PR curves drawn per variant, with operating points marked.
- [ ] Pareto (security vs latency) figure generated and interpretation paragraph drafted.
- [ ] Weak/non-contributing components identified; supervisor decides demotion vs claim.
- [ ] All figures/tables generated from raw CSVs via reproducible script; paper-ready markdown table drafts produced.

---

## Handoff to Next Issue

**HG6 — Statistical Validation & Reproducibility Artifact** mirrors KV6: k-fold repeats, bootstrap CIs, McNemar, regenerate paper numbers from raw CSVs with one click, environment pinned.
